fix: improve body parsing logic

Ansuel opened the pull request:
Current body parsing logic with trim() + split("###") is too fragile and
pose problems with some body that contains case with ### in the middle
of the line or case with codeblock ``` section.

These case will cause the script to parse these as separate section and
produce wrong outputs and in some case even prints error assuming things
are checkbox and errors out on the concat function.

To make the parsing logic more solid, implement a dedicated function and
parse with this logic:
- We split the body for "\n"
- We ignore codeblock ``` section
- We check "###" only at the start of the line
- We check for "### " (with the space included) as that is the correct
  section Github issue template expects.

With the following change case like:
```
root@OpenWrt:~# cat /boot/config.txt
```

Are correctly parsed as all part of a single section instead of being
wrongly treated as different empty sections.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
feat: improve body parsing logic

Current body parsing logic with trim() + split("###") is too fragile and
pose problems with some body that contains case with ### in the middle
of the line or case with codeblock ``` section.

These case will cause the script to parse these as separate section and
produce wrong outputs and in some case even prints error assuming things
are checkbox and errors out on the concat function.

To make the parsing logic more solid, implement a dedicated function and
parse with this logic:
- We split the body for "\n"
- We ignore codeblock ``` section
- We check "###" only at the start of the line
- We check for "### " (with the space included) as that is the correct
  section Github issue template expects.

With the following change case like:
```
root@OpenWrt:~# cat /boot/config.txt
```

Are correctly parsed as all part of a single section instead of being
wrongly treated as different empty sections.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
feat: add additional test for section with ####

Add additional test for section with #### to prevent and catch in
future code change that would produce wrong parsing.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
feat: add additional test for codeblock ``` ignore

Add test for codeblock ``` ignore to prevent and catch in future code
change that would produce wrong parsing.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
Discovered by this https://github.com/openwrt/openwrt/issues/15265

Made further research and found the root of the problem!

jamacku commented on index.js:
````diff
@@ -111,9 +111,35 @@ async function run(env, body, fs, core) {
+
+    body.split("\n").reduce((str, line, idx, arr) => {
+      // ``` Section must be ignored and not parsed as new section
+      if (line == "```")
````
Please also account for instances when the context of the code block is specified. For example:

```
```md
```

Ansuel replied:
Funnily enough I just encounter this bug and was ready to report this problem just only to discover I was the culprit LOL...

my bad...

Ansuel committed:
feat: add additional test for codeblock ```sh ignore

Add test for codeblock ```sh ignore to prevent and catch in future code
change that would produce wrong parsing.

This is a variant of the ``` test that makes the code block section
target specific code.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@jamacku sorry for the delay... I fixed the comment and also add an additional test to catch this special case. We check now if line start with ``` directly.

jamacku reviewed (approved):
LGTM, Thank you.

stefanbuck commented:
Awesome, good stuff. I will have a closer look shortly and then cut a new major version for this change.

gr2m committed:
test: convert new fixtures and tests to ESM

The fixtures and test blocks added in this PR used CommonJS
(require / module.exports). After rebasing onto main, which
converted the codebase to ESM in #105, they need to use
import / export default and the loadJson helper.

gr2m commented:
Gonna merge as fix. No breaking changes are introduced. Thanks @Ansuel!

Ansuel commented:
@gr2m thanks a lot!
